openmeter/ent/schema/productcatalog.go

```diff
@@ -0,0 +1,251 @@
+	"github.com/openmeterio/openmeter/pkg/datex"
+	"github.com/openmeterio/openmeter/pkg/framework/entutils"
+)
+
+type Plan struct {
+	ent.Schema
+}
+
+func (Plan) Mixin() []ent.Mixin {
+	return []ent.Mixin{
+		entutils.IDMixin{},
+		entutils.KeyMixin{},
+		entutils.NamespaceMixin{},
+		entutils.TimeMixin{},
+		entutils.MetadataAnnotationsMixin{},
+	}
+}
+
+func (Plan) Fields() []ent.Field {
+	return []ent.Field{
+		field.Int("version").
+			Min(1),
+		field.String("name").
+			NotEmpty(),
+		field.String("description").
+			Optional().
+			Nillable(),
+		field.String("currency").
+			Default("USD").
+			NotEmpty().
+			Immutable(),
+		field.Time("effectiveFrom").
+			Optional().
+			Nillable(),
+		field.Time("effectiveTo").
+			Optional().
+			Nillable(),
+	}
+}
+
+func (Plan) Edges() []ent.Edge {
+	return []ent.Edge{
+		edge.To("phases", PlanPhase.Type).
+			Annotations(entsql.Annotation{
+				OnDelete: entsql.Cascade,
+			}),
+	}
+}
+
+func (Plan) Indexes() []ent.Index {
+	return []ent.Index{
+		index.Fields("namespace", "id"),
+		index.Fields("namespace", "key", "version").
+			Unique(),
+	}
+}
+
+type PlanPhase struct {
+	ent.Schema
+}
+
+func (PlanPhase) Mixin() []ent.Mixin {
+	return []ent.Mixin{
+		entutils.IDMixin{},
+		entutils.KeyMixin{},
+		entutils.NamespaceMixin{},
+		entutils.TimeMixin{},
+		entutils.MetadataAnnotationsMixin{},
+	}
+}
+
+func (PlanPhase) Fields() []ent.Field {
+	return []ent.Field{
+		field.String("start_after").
+			GoType(datex.ISOString("")).
+			Default(plan.DefaultStartAfter),
+		field.String("name").
+			NotEmpty(),
+		field.String("description").
+			Optional().
+			Nillable(),
+		field.String("discounts").
+			GoType([]plan.Discount{}).
+			ValueScanner(DiscountsValueScanner).
+			SchemaType(map[string]string{
+				dialect.Postgres: "jsonb",
+			}).
+			Optional().
+			Nillable(),
+		field.String("plan_id").
+			NotEmpty().
+			Comment("The plan identifier the phase is assigned to."),
+	}
+}
+
+func (PlanPhase) Edges() []ent.Edge {
+	return []ent.Edge{
+		edge.From("plan", Plan.Type).
+			Ref("phases").
+			Field("plan_id").
+			Required().
+			Unique(),
+		edge.To("ratecards", PlanRateCard.Type).
+			Annotations(entsql.Annotation{
+				OnDelete: entsql.Cascade,
+			}),
+	}
+}
+
+func (PlanPhase) Indexes() []ent.Index {
+	return []ent.Index{
+		index.Fields("namespace", "key"),
+		index.Fields("plan_id", "key").
+			Unique(),
+	}
+}
+
+type PlanRateCard struct {
+	ent.Schema
+}
+
+func (PlanRateCard) Mixin() []ent.Mixin {
+	return []ent.Mixin{
+		entutils.IDMixin{},
+		entutils.KeyMixin{},
+		entutils.NamespaceMixin{},
+		entutils.TimeMixin{},
+		entutils.MetadataAnnotationsMixin{},
+	}
+}
+
+type Price string
+
+func (PlanRateCard) Fields() []ent.Field {
+	return []ent.Field{
+		field.Enum("type").
+			GoType(plan.RateCardType("")).
+			Immutable(),
+		field.String("name").
+			NotEmpty(),
+		field.String("description").
+			Optional().
+			Nillable(),
+		field.String("feature_key").
```

Comment:
these are mutable because user can edit plan in draft, right?

Reply:
Yup, thats the idea.